modules/phase_field/include/kernels/MatDiffusionBase.h

```diff
@@ -0,0 +1,59 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef MATDIFFUSIONBASE_H
+#define MATDIFFUSIONBASE_H
+
+#include "DerivativeMaterialInterface.h"
+#include "Kernel.h"
+
+template<typename T>
+class MatDiffusionBase : public DerivativeMaterialInterface<Kernel>
+{
+public:
+  MatDiffusionBase(const InputParameters & parameters);
+
+  static InputParameters validParams();
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+
+  const MaterialProperty<T> & _D;
+};
+
+template<typename T>
+InputParameters
+MatDiffusionBase<T>::validParams()
+{
+  InputParameters params = ::validParams<Kernel>();
+  params.addClassDescription("Diffusion equation Kernel that takes teh Diffusivity from a material property");
+  params.addParam<MaterialPropertyName>("D_name", "D", "The name of the diffusivity");
+  return params;
+}
+
+template<typename T>
+MatDiffusionBase<T>::MatDiffusionBase(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Kernel>(parameters),
+    _D(getMaterialProperty<T>("D_name"))
+{
+}
+
+template<typename T>
+Real
+MatDiffusionBase<T>::computeQpResidual()
+{
+  return _D[this->_qp] * this->_grad_test[this->_i][this->_qp] * this->_grad_u[this->_qp];
```

Comment:
What's with all the "this" dereferencing? Is it necessary because this is a template class?